src/main/java/org/mule/extension/http/api/listener/headers/InvalidTransferEncodingValidator.java

```diff
@@ -0,0 +1,142 @@
+
+  private static final Logger LOGGER = getLogger(InvalidTransferEncodingValidator.class);
+  private static final String TRANSFER_ENCODING_LOWERCASE = TRANSFER_ENCODING.toLowerCase(Locale.ROOT);
+  private static final int ERRORS_LIMIT_TO_PRINT_WARNING = 2 << 20;
+  private final boolean throwException;
+  private final AtomicInteger errorsFound = new AtomicInteger();
+
+  /**
+   * @param throwException if {@code true}, the errors will be reported as an exception. Otherwise, the errors will be
+   *                       logged. One of each {@link ERRORS_LIMIT_TO_PRINT_WARNING} errors will be printed as a WARN log,
+   *                       and the others as TRACE, in order to avoid log flooding.
+   */
+  public InvalidTransferEncodingValidator(boolean throwException) {
+    this.throwException = throwException;
+  }
+
+  /**
+   * Checks that 'Transfer-Encoding' header value has a valid value.
+   *
+   * @param headers all the headers in the request.
+   * @throws HttpHeadersException with a 400 (Bad Request) if the 'Transfer-Encoding' value is invalid.
+   */
+  @Override
+  public void validateHeaders(MultiMap<String, String> headers) throws HttpHeadersException {
+    List<String> allTransferEncodings = headers.getAll(TRANSFER_ENCODING_LOWERCASE);
+    int numberOfHeaders = allTransferEncodings.size();
+
+    // avoid creating the implicit iterator in order to optimize performance
+    if (numberOfHeaders == 0) {
+      return;
+    }
+    if (numberOfHeaders == 1) {
+      if (!isValidTransferEncodingHeader(allTransferEncodings.get(0))) {
+        reportError();
+      }
+      return;
+    }
+
+    for (String header : allTransferEncodings) {
+      if (!isValidTransferEncodingHeader(header)) {
+        reportError();
+      }
+    }
+  }
+
+  private void reportError() throws HttpHeadersException {
+    if (throwException) {
+      throw new HttpHeadersException(format("'%s' header has an invalid value", TRANSFER_ENCODING), BAD_REQUEST);
+    } else {
+      if (errorsFound.getAndIncrement() % ERRORS_LIMIT_TO_PRINT_WARNING == 0) {
+        LOGGER.warn("'{}' header has an invalid value", TRANSFER_ENCODING);
+      } else {
+        LOGGER.trace("'{}' header has an invalid value", TRANSFER_ENCODING);
+      }
+    }
+  }
+
+  private static boolean isValidTransferEncodingHeader(String headerValue) {
+    if ("chunked".equals(headerValue)) {
+      // optimize the common case
+      return true;
+    }
+
+    if (headerValue.contains(",")) {
+      for (String singleTransferEncoding : headerValue.split(",")) {
+        if (!isSingleHeaderValid(singleTransferEncoding)) {
+          return false;
+        }
+      }
+      return true;
+    } else {
+      return isSingleHeaderValid(headerValue);
+    }
+  }
+
+  private static boolean isSingleHeaderValid(String transferEncoding) {
+    String trimmed = trimIfNeeded(transferEncoding);
+    int size = trimmed.length();
+    switch (size) {
+      case 7:
```

Comment:
precalculate the sized of the known headers as constants, so it is understood what they represent instead of being magic numbers

Reply:
I didn't calculate it because it must be a constant expression. In the feedback I extracted these numbers to constants